### PR TITLE
Fix K8S executor override config using pod_override_object

### PR DESCRIPTION
**Pull request by hussein-awala**
related: #22298

Currently, we don't use all the configs from `pod_override_object` where we override some of them by the default K8S executor configurations or the kube config file (ex: default namespace).

This PR fixes the issue by changing the order of `pod_override_object` in the configs source list to use all the configurations provided in this object.

### CODE REVIEW

**Review by eladkal — APPROVED** [review]
Thanks for this fix!

**Comment by eladkal** [comment]
I am not sure why helm chart tests are skipped here cc @potiuk 
In previous attempt to address the issue https://github.com/apache/airflow/pull/34505 we had failing helm tests so I will add full tests labels just to make sure this change is OK

**Comment by eladkal** [comment]
Helm tests are failing :(

**Comment by hussein-awala** [comment]
I'm still trying to reproduce the tests failure locally:
```
test_kubernetes_executor.py::TestKubernetesExecutor::test_integration_run_dag PASSED                                                                                     [  1%]
test_kubernetes_executor.py::TestKubernetesExecutor::test_integration_run_dag_with_scheduler_failure PASSED                                                              [  3%]
test_kubernetes_pod_operator.py::TestKubernetesPodOperatorSystem::test_do_xcom_push_defaults_false PASSED                                                                [  5%]
test_kubernetes_pod_operator.py::TestKubernetesPodOperatorSystem::test_config_path_move PASSED                                                                          
````

**Comment by hussein-awala** [comment]
@potiuk, do you have an idea why the K8S executor tests failed? I tested locally with Mac arm and Linux amd, the tests finished successfully.

**Comment by potiuk** [comment]
This branch is [4 commits ahead](https://github.com/hussein-awala/airflow/compare/apache:airflow:main...fix/k8s_executor_ns_override), [18 commits behind](https://github.com/hussein-awala/airflow/compare/fix/k8s_executor_ns_override...apache:airflow:main) apache:main.


I just rebased it. I think https://github.com/apache/airflow/pull/35191 should fix it.

**Comment by potiuk** [comment]
But let's see.

**Comment by potiuk** [comment]
I guess changing timeout will not work here. This looks like a real problem introduced by the change  - k8s executor stops working and it should be investigated. It's rather easy to reproduce all that CI does here locally and it's well described in https://github.com/apache/airflow/blob/main/TESTING.rst#typical-testing-pattern-for-kubernetes-tests 

Following the steps describing there will setup precisely the same environment, kind cluster, will build airflow and deploy the image with airflow and test dags + test environment where the tests will interact with the cluster and (for some reason) fail.

**Comment by hussein-awala** [comment]
> I guess changing timeout will not work here. This looks like a real problem introduced by the change - k8s executor stops working and it should be investigated. It's rather easy to reproduce all that CI does here locally and it's well described in https://github.com/apache/airflow/blob/main/TESTING.rst#typical-testing-pattern-for-kubernetes-tests
> 
> Following the steps describing there will setup precisely the same environment, kind cluster, will build airflow and deploy the image with airflow and test dags + test environment where the tests will interact with the cluster and (for some reason) fail.

I already reproduced all the steps on two different computers step by step, and I also tested `run-complete-tests`, in both cases, the tests finished successfully.

I also triggered the dag manually via the UI deployed in the kind cluster, and all worked as expected.

**Comment by potiuk** [comment]
FYI. I removed the "full tests needed" and rebased to check if there is indeed an issue that "K8S System tests" are not triggered for such PR - I think they should be - looking at selective checks output and the code, but let's see (cc: @eladkal )

**Comment by potiuk** [comment]
> I already reproduced all the steps on two different computers step by step, and I also tested `run-complete-tests`, in both cases, the tests finished successfully.

Oh I missed that. Maybe the logs will help ?  If you look at summary https://github.com/apache/airflow/actions/runs/6679106455  and scroll down a bit - you will find that there are complete logs (dumps from kind logs) available as artifacts - they should contain more details on what's going on during the tests.

**Comment by potiuk** [comment]
BTW. Those tests are pretty stable in general, so it's quite for sure some effect of those changes.

**Comment by eladkal** [comment]
@hussein-awala were you able to find some info about the problem? I tried also to check the logs but couldn't find the root cause for the failures

**Comment by jedcunningham** [comment]
My best guess is that this is what is failing: https://github.com/apache/airflow/blob/9356233762a67c0a787fc3e16040455019af13bd/airflow/example_dags/example_kubernetes_executor.py#L152-L160

I see nothing in the repo that is provisioning that namespace in kind.

If we merge this, we need to update this doc:
[https://github.com/apache/airflow/blob/9356233762a67c0a787fc3e16040455019af13bd/do[…]cs/apache-airflow/core-concepts/executor/kubernetes.rst?plain=1](https://github.com/apache/airflow/blob/9356233762a67c0a787fc3e16040455019af13bd/docs/apache-airflow/core-concepts/executor/kubernetes.rst?plain=1#L137)


I think it’d be worth adding more coverage to [this test](https://github.com/apache/airflow/blob/9356233762a67c0a787fc3e16040455019af13bd/tests/providers/cncf/kubernetes/test_pod_generator.py#L342) to make sure things like the dag_id etc that Airflow sets wins.

**Comment by hussein-awala** [comment]
> My best guess is that this is what is failing:
> 
> https://github.com/apache/airflow/blob/9356233762a67c0a787fc3e16040455019af13bd/airflow/example_dags/example_kubernetes_executor.py#L152-L160
> 
> I see nothing in the repo that is provisioning that namespace in kind.
> 
> If we merge this, we need to update this doc: [https://github.com/apache/airflow/blob/9356233762a67c0a787fc3e16040455019af13bd/do[…]cs/apache-airflow/core-concepts/executor/kubernetes.rst?plain=1](https://github.com/apache/airflow/blob/9356233762a67c0a787fc3e16040455019af13bd/docs/apache-airflow/core-concepts/executor/kubernetes.rst?plain=1#L137)
> 
> I think it’d be worth adding more coverage to [this test](https://github.com/apache/airflow/blob/9356233762a67c0a787fc3e16040455019af13bd/tests/providers/cncf/kubernetes/test_pod_generator.py#L342) to make sure things like the dag_id etc that Airflow sets wins.

@jedcunningham When I ran the test locally with breeze, all the dag tasks (including `other_namespace_task`, which you mentioned) finished successfully, I even added `--multi-namespace-mode` to breeze k8s to add multi-namespace configurations.

**Comment by potiuk** [comment]
Can you reabase again/resolve conflicts please ? The artifacts have expired now - so we cannot even see what happened in the logs of K8s here.

**Comment by potiuk** [comment]
The artifacts containing log dumps from kind have more information on the root cause

If you download the artifacts and look at the logs you will find this (in one of the scheduler logs):

```
2023-11-10T19:30:51.588963789Z stdout F [2023-11-10T19:30:51.588+0000] {rest.py:231} DEBUG - response body: {"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"pods \"example-kubernetes-executor-other-namespace-task-uhm9sfrj\" is forbidden: error looking up service account test-namespace/airflow-worker: serviceaccount \"airflow-worker\" not found","reason":"Forbidden","details":{"name":"example-kubernetes-executor-other-namespace-task-uhm9sfrj","kind":"pods"},"code":403}
2023-11-10T19:30:51.588970921Z stdout F 
```

There are some more details available (details of "Pod Creation Request" that failed:

```
2023-11-10T19:30:51.584086049Z stdout F [2023-11-10T19:30:51.583+0000] {kubernetes_executor_utils.py:309} DEBUG - Pod Creation Request: 
2023-11-10T19:30:51.584092635Z stdout F {
2023-11-10T19:30:51.58409715Z stdout F   "apiVersion": "v1",
2023-11-10T19:30:51.584099589Z stdout F   "kind": "Pod",
2023-11-10T19:30:51.584101837Z stdout F   "metadata": {
2023-11-10T19:30:51.584103846Z stdout F     "annotations": {

````

**Comment by potiuk** [comment]
Just go to "Summary" and scroll down to see the artifacts https://github.com/apache/airflow/actions/runs/6827578309?pr=35185

**Comment by hussein-awala** [comment]
Thanks @potiuk! I will check why I don't have the same issue in local.

I will update the breeze command to create this serviceaccount
> error looking up service account test-namespace/airflow-worker: serviceaccount \"airflow-worker\" not found

**Comment by potiuk** [comment]
> Thanks @potiuk! I will check why I don't have the same issue in local.
> 
> I will update the breeze command to create this serviceaccount
> 
> > error looking up service account test-namespace/airflow-worker: serviceaccount "airflow-worker" not found

Yeah. Would be great to find out the root cause of `CI <-> local` - they should be IN THEORY the same.. But you know:

> In theory practice is the same as theory, but in practice it isn't

**Comment by potiuk** [comment]
And BTW. My wild guess - without looking at the changes/code is that this is some kind of race condition - where the service account is either not created yet or maybe while being deleting and recreated when it is needed.

The tests we run in CI are (quite a bit deliberately)  running in a rather constrained environment - we run up to 3 parallel Kind clusters and 3 parallel clusters at the same time - this means that we are stretching the limits of the machine it runs on - it has 8 cores and 64 GB memory, but it still has a lot of I/O  contention and it might simply run a lot slower at times when a lot of other things happen - and that's  not only good to  squeeze everything out of that machine (and get the test finish as soon as possible) but also this is ideal environment where those kinds of race conditions happen - this is why our flaky tests are more flaky in CI usually, because the resources for single run are limited. 

This way we can also detect REAL issues - that we would only find out when our users try to run Airflow on a "minimum hardware" they pay as little as possible for. And this error smells like one of those cases where if you try to install helm in such limited resource environment (with these namespace modifications) it might fall into one of those race conditions.

**Comment by hussein-awala** [comment]
K8S tests are green, I will move the helm chart changes to a separate PR and provide a small patch for these tests instead.

**Comment by potiuk** [comment]
Oh. What was it?

File: dev/breeze/src/airflow_breeze/commands/kubernetes_commands.py [hunk 1]
```diff
@@ -1036,8 +1047,47 @@ def _deploy_airflow(
...
+                f"kubectl get serviceaccount -n {HELM_AIRFLOW_NAMESPACE} "
+                "--field-selector  metadata.name!=default -o yaml "
+                f"| sed 's/namespace: {HELM_AIRFLOW_NAMESPACE}/namespace: {TEST_NAMESPACE}/' "
+                f"| kubectl apply -n {TEST_NAMESPACE} -f -",
+                python=python,
+                kubernetes_version=kubernetes_version,
+                output=output,
+                check=False,
+                shell=True,
+            )
```
Comment (hussein-awala) [inline dev/breeze/src/airflow_breeze/commands/kubernetes_commands.py:1089]: These commands will duplicate the resources used in pod_template to test-namespace:
```
secret/airflow-broker-url created
secret/airflow-fernet-key created
secret/airflow-metadata created
secret/airflow-postgresql created
secret/airflow-redis-password created
secret/airflow-webserver-secret-key created
configmap/airflow-config created
configmap/airflow-statsd created
serviceaccount/airflow-create-user-job created
serviceaccount/airflow-migrate-database-job created
serviceaccount/airflow-scheduler created
serviceaccount/airflow-statsd created
serviceaccount/airflow-triggerer created
serviceaccount/airflow-webserver created
serviceaccount/airflow-worker created
```
I excluded the helm secrets/configmaps and the default service account from the duplication operation.

**Review by potiuk — APPROVED** [review]
NAAAJS

**Comment by potiuk** [comment]
I guess later we can add multi-namespace mode to run in CI. We likely do not want to add another dimention for k8s tests but we can do what we do in databases with some kind of rotating scheme of test combos (k8s version, standard-naming, multi-namespace) - but it can be added later.

**Comment by hussein-awala** [comment]
> I guess later we can add multi-namespace mode to run in CI. We likely do not want to add another dimention for k8s tests but we can do what we do in databases with some kind of rotating scheme of test combos (k8s version, standard-naming, multi-namespace) - but it can be added later.

I agree, and it will be much simpler with https://github.com/apache/airflow/pull/35639, I hope finding some time to finish it before the next chart release.

**Comment by hussein-awala** [comment]
Finally 🎉

**Pull request merged**